docs/RFCS/20220711_external_connections.md

```diff
@@ -0,0 +1,407 @@
+  // ConnectionType returns the type of the External Connection.
+  ConnectionType() connectionpb.ConnectionType
+  // ConnectionProto prepares the ConnectionDetails for serialization.
+  ConnectionProto() *connectionpb.ExternalConnectionDetails
+  // Dial establishes a connection to the external resource.
+  Dial() Connection 
+  // DisplayString returns the string representation of the
+  // external resource with all sensitive information redacted.
+  DisplayString() string
+}
+
+// Connection is a marker interface that is implemented by objects that
+// support interaction with the external resource.  
+type Connection interface {}
+```
+
+The `ConnectionProto` method returns the concrete `connectionpb.ExternalConnectionDetails` that can
+be serialized and stored in the `connection_details` columns of the system table.
+The `ConnectionType` method returns the connection type that the External Connection belongs to.
+
+To prevent misuse of the underlying endpoint URI and configuration information, we will expose an
+opaque `Dial` method. This method will be responsible for establishing a connection to the external
+resource and returning a handle to interact with it. The `Connection` interface is a marker
+interface that will be implemented by the `cloud.ExternalStorage`, `cloud.KMS`,
+and `changefeedccl.Sink` interfaces described in the section below. The handle will therefore
+support all the methods that are used by bulk operations, and changefeeds to interact with external
+connections today.
+
+The `DisplayString` method returns a string representation of the external connection with all
+secrets redacted. This string will be used by observability tools that display information about the
+external resource. The scheme-specific implementations have control over how to display their
+endpoint URI and configurations as well as what information they consider privileged.
+
+### Accessing an External Connection
+
+Database operations such as backup, restore, import, export, and changefeeds interact with external
+resources via user-inputted URIs as part of the SQL query. Depending on the database operation and
+the scheme of the URI, the parsing, validation, and creation of a handle to the external connection
+are managed by different packages in CockroachDB. Each package recognizes a set of URI schemes and
+in turn returns one of the following interfaces to interact with the external connection:
+
+`cloud.ExternalStorage`: Used by bulk operations to read and write files.
+
+`cloud.KMS`: Used by bulk operations to encrypt and decrypt data using a Key Management Service.
+
+`changefeedccl.Sink`: Used by changefeeds to stream changes to an external connection.
+
+This RFC proposes introducing a new URI scheme that can be used with all bulk operations and
+changefeeds. The URI will be of the form `external://<external-connection-name>`.
+
+The URI will only permit a hostname, in other words, it will not accept a path or any parameters.
```

Comment:
I think we could allow a Path component to be present and if it is, we just append it to the named connection's preset path while making the external storage?

Reply:
@rhu713 had mentioned this in the google doc from the point of view of a user wanting to specify an `incremental_location` for backups in the same bucket but a different subdirectory. My first reaction was that we are not making the UX any worse than it is today by having the user specify two unique `external` locations in the backup statement. 

By accepting a path, we are once again tying the representation of the external connection with a particular use case. External Storage is one of the resources that an External Connection can represent. I'm wary of having path mean something for Storage, something else for KMS, and so on. In the future, if we want to phase out the URI syntax entirely, and only accept a connection name, then adding a path complicates this too?

Reply:
I'm starting to think that we might need to do this sooner rather than later. In a lot of places in our code we construct ExternalStorage objects using URIs that are original URI passed in as part of the SQL query + some suffix. For example, if we are to run `BACKUP INTO 'external://foo`, during backup destination resolution we actually construct an ExternalStorage to `external://foo/<some-timestamp-path>`. So if we are rejecting the `path` component of an `external` URI then we need to mandate every user to construct an `ExternalStorage` to the base URI (represented by the external connection object), and then pass in the relevant prefix while reading or writing objects. 

I don't love the idea of having an s3/gcs subdirectory appended to `external://foo` but it seems less of an uphill battle than preventing any user of `MakeExternalStorageFromURI` from passing in a path component. Furthermore in workflows such as `SHOW BACKUPS IN 'external://foo'` followed by `SHOW BACKUP 'external://foo/<particular-collection>'`, it might make for a better UX as you mentioned before.